File: machine/cpp/machine/devices/input/sample_edge.cpp
#include "machine/devices/input/sample_edge.h"

#include "machine/bus/io.h"
#include "machine/devices/input/registers.h"
#include "machine/devices/input/sample_latch.h"
#include "machine/memory/memory.h"

namespace bmsx {

InputControllerSampleEdge::InputControllerSampleEdge(InputControllerInputSource& input, InputControllerSampleLatch& sampleLatch, InputControllerRegisterFile& registers, Memory& memory)
	: m_input(input)
	, m_sampleLatch(sampleLatch)
	, m_registers(registers)
	, m_memory(memory) {}

void InputControllerSampleEdge::onVblankEdge(f64 currentTimeMs, u32 nowCycles) {
	if (!m_sampleLatch.consumeVblankEdge(nowCycles)) {
		return;
	}
	m_input.sampleInputControllerSnapshot(currentTimeMs, m_snapshot);
	m_registers.latchSnapshot(m_snapshot);
	m_registers.mirror(m_memory);
	m_memory.writeIoValue(IO_INP_STATUS, valueNumber(static_cast<double>(m_sampleLatch.sequence())));
}

} // namespace bmsx

File: machine/ts/machine/devices/input/sample_edge.ts
import { IO_INP_STATUS } from '../../bus/io';
import { Memory } from '../../memory/memory';
import type { InputControllerSampleLatch } from './sample_latch';
import type { InputControllerRegisterFile } from './registers';
import { createInputControllerSnapshot, type InputControllerInputSource } from './contracts';

export class InputControllerSampleEdge {
	private readonly snapshot = createInputControllerSnapshot();

	public constructor(
		private readonly input: InputControllerInputSource,
		private readonly sampleLatch: InputControllerSampleLatch,
		private readonly registers: InputControllerRegisterFile,
		private readonly memory: Memory,
	) {}

	public onVblankEdge(currentTimeMs: number, nowCycles: number): void {
		if (!this.sampleLatch.consumeVblankEdge(nowCycles)) {
			return;
		}
		this.input.sampleInputControllerSnapshot(currentTimeMs, this.snapshot);
		this.registers.latchSnapshot(this.snapshot);
		this.registers.mirror(this.memory);
		this.memory.writeIoValue(IO_INP_STATUS, this.sampleLatch.sequence());
	}
}
